### DISCUSSION TITLE
FastAPI - need to asyncio.sleep() or event is not even emitted

### DISCUSSION THREAD

**Discussion by seatedscribe** [miguelgrinberg/python-socketio#1093] [Q&A] [answered]
I am trying to implement a logic similar to [this](https://stackoverflow.com/questions/43301977/flask-socket-io-result-of-emit-callback-is-the-response-of-my-rest-endpoint/) using async client and fastAPI+socketio server.

After a lot of try/fail I found that an `asyncio.sleep(.1)` is needed or the callback is never executed (actually it seems that `emit`ted event does not even reach the other side). See comments near the end of the code body.

Cannot understand what is it going on here, any help appreciated.

```
# server.py
# invoke command
# python -m uvicorn server:app --reload --port 8765


from fastapi import FastAPI
import socketio
import asyncio
import threading
import logging
import sys

FORMAT = '[%(asctime)s] %(message)s'

logging.basicConfig(level=logging.DEBUG,
                    format=FORMAT,
                    stream=sys.stdout)
_LOGGER = logging.getLogger(__name__)


sio = socketio.AsyncServer(
    async_mode='asgi',
    cors_allowed_origins='*',
    engineio_logger=True
)

fastapi = FastAPI()

app = socketio.ASGIApp(
    socketio_server=sio,
    other_asgi_app=fastapi,
    socketio_path='/socket.io/'
)

SID = ""


@sio.event
def connect(sid, environ):
    global SID
    SID = sid
    _LOGGER.info(f"+++ a client just connected: {SID}")


@fastapi.get('/test_wait')
async def wait():
    global SID
    ev = threading.Event()
    result = None

    def mycallback(data):
        nonlocal result
        nonlocal ev

        _LOGGER.info(f"CALLBACK {data}")
        ev.set()  # unblock HTTP route
        _LOGGER.info(f"---> EVENT SET")
        result = data

    await sio.emit(event='test_event', data='dummy data', room=SID, callback=mycallback)
    _LOGGER.info(f"event emitted. waiting...")
    #############
    # without asyncio sleeping, ev.wait never returns because emitted event did not even reach the client
    #############
    await asyncio.sleep(.1)
    ev.wait()
    return result
```

(complete code/requirements and log dump available on request)

**Comment by miguelgrinberg** [comment, marked as answer]
You are using a `threading.Event` with asyncio, which does not really make sense, since it is blocking.  The `ev.wait()` is going to block the asyncio loop, preventing all the background activity from Socket.IO from happening (along with everything else that goes on in your application). The sleep gives you the impression that it addresses the problem only because during that 0.1 of a second the background work needed by the emit is completed. But you are still blocking the loop right after.

I don't understand why are you mixing a threading.Event in your async application. Asyncio has event objects, why not use those?

**Reply by seatedscribe** [reply]
Duh! Of course it had to be some sort of gross mistake, but sometimes the elephant in front of me has invisible powers.
Thanks for your time!
